### PR TITLE
Retain information about narrowed type when doing list

### PR DESCRIPTION
**Pull request by jgogstad**
* Retain knowledge about the narrowed type when doing `.list`
* Introduce SftpPath for capturing sftp path attributes
* Connect Sftp `Session` if client haven't connected it.

### CODE REVIEW

File: sftp/src/main/scala/blobstore/sftp/SftpStore.scala [hunk 1]
```diff
@@ -41,6 +40,8 @@ final class SftpStore[F[_]](
 )(implicit F: ConcurrentEffect[F], CS: ContextShift[F])
   extends Store[F] {
 
+  override type NarrowedPath = Path
```
Comment (gafiatulin) [inline sftp/src/main/scala/blobstore/sftp/SftpStore.scala:43]:
```suggestion
  override type NarrowedPath = SftpPath
```

File: core/src/main/scala/blobstore/StoreOps.scala [hunk 2]
```diff
@@ -109,14 +109,14 @@ trait StoreOps {
 
   }
 
-  implicit class ListOps[F[_]](store: Store[F]) {
+  implicit class ListOps[F[_]](val store: Store[F]) {
```
Comment (gafiatulin) [inline core/src/main/scala/blobstore/StoreOps.scala:112]: Any reason for making `store` a val here?

File: core/src/main/scala/blobstore/fs/FileStore.scala [hunk 3]
```diff
@@ -25,6 +25,9 @@ import fs2.{Pipe, Stream}
 
 final class FileStore[F[_]](fsroot: NioPath, blocker: Blocker)(implicit F: Sync[F], CS: ContextShift[F])
   extends Store[F] {
+
+  override type NarrowedPath = Path
```
Comment (gafiatulin) [inline core/src/main/scala/blobstore/fs/FileStore.scala:29]: Maybe can be `BasePath` as well?

File: sftp/src/main/scala/blobstore/sftp/SftpPath.scala [hunk 4]
```diff
@@ -0,0 +1,17 @@
+package blobstore.sftp
+
+import java.time.Instant
+
+import blobstore.Path
+import cats.data.Chain
+import com.jcraft.jsch.SftpATTRS
+
+case class SftpPath(root: Option[String], fileName: Option[String], pathFromRoot: Chain[String], attributes: SftpATTRS)
```
Comment (gafiatulin) [inline sftp/src/main/scala/blobstore/sftp/SftpPath.scala:9]: Nice! Wasn't sure if `SftpATTRS` brings anything useful except 3 already exposed attributes.

File: core/src/main/scala/blobstore/StoreOps.scala (hunk 2 above)
Comment (gafiatulin) [reply core/src/main/scala/blobstore/StoreOps.scala:112]: >private value store escapes its defining scope as part of type F[List[ListOps.this.store.NarrowedPath]]

I see, never mind.

**Comment by gafiatulin** [comment]
Any situation where Path-Dependent Types can lead to some problems?

File: sftp/src/main/scala/blobstore/sftp/SftpPath.scala (hunk 4 above)
Comment (jgogstad) [reply sftp/src/main/scala/blobstore/sftp/SftpPath.scala:9]: I think `uid`/`gid` could be useful for filtering, doesn't hurt to expose creation date either. We could maybe map all these fields, not just a subset of them. I'll do it in another PR.

**Comment by jgogstad** [comment]
> Any situation where Path-Dependent Types can lead to some problems?

No, but I figured that they're not necessary. `fs2.Stream` is covariant in it's second parameter, so there's no need to retain the type like this. Users probably have a `S3Store` to begin with, so just changing the type in `list` to be `S3Path` etc. achieves the same thing without the type member.

Changing to that now. I reckon this changeset will be a lot smaller.

**Pull request merged**